### PR TITLE
OZ-L06: try-catch permit2Forwarder to avoid exposing DOS on multicall

### PR DESCRIPTION
**Pull request by saucepoint**
## Related Issue
OZ-L06: allow permit2Forwarder's `permit`/`permitBatch` to fail via try-catch.

Prevents DOS'ing multicalls containing `permit`/`permitBatch` where an external party front runs the permit2Forwarder call

## Description of changes

try-catch'd permit2Forwarder's `permit`/`permitBatch`

### CODE REVIEW

File: src/base/Permit2Forwarder.sol [hunk 1]
```diff
@@ -27,6 +27,6 @@ contract Permit2Forwarder {
         external
         payable
     {
-        permit2.permit(owner, _permitBatch, signature);
+        try permit2.permit(owner, _permitBatch, signature) {} catch {}
```
Comment (hensha256) [inline src/base/Permit2Forwarder.sol:30]: maybe a quick comment that explains why its in a try catch?

File: test/position-managers/PositionManager.multicall.t.sol [hunk 2]
```diff
@@ -321,4 +331,76 @@ contract PositionManagerMulticallTest is Test, Permit2SignatureHelpers, PosmTest
...
+            defaultERC20PermitAllowance(Currency.unwrap(currency1), permitAmount, permitExpiration, permitNonce);
+        permit1.spender = address(lpm);
+        bytes memory sig1 = getPermitSignature(permit1, charliePK, PERMIT2_DOMAIN_SEPARATOR);
+
+        // bob front-runs the permits
+        vm.startPrank(bob);
+        lpm.permit(charlie, permit0, sig0);
+        lpm.permit(charlie, permit1, sig1);
+        vm.stopPrank();
+
```
Comment (hensha256) [inline test/position-managers/PositionManager.multicall.t.sol:359]: check that the permit has succeeded by checking permit2 has stored the permitNonce has been taken i think?

File: test/position-managers/PositionManager.multicall.t.sol [hunk 3]
```diff
@@ -321,4 +331,76 @@ contract PositionManagerMulticallTest is Test, Permit2SignatureHelpers, PosmTest
...
+        // charlie tries to mint an LP token with multicall(permit, permit, mint)
+        bytes[] memory calls = new bytes[](3);
+        calls[0] = abi.encodeWithSelector(Permit2Forwarder(lpm).permit.selector, charlie, permit0, sig0);
+        calls[1] = abi.encodeWithSelector(Permit2Forwarder(lpm).permit.selector, charlie, permit1, sig1);
+        bytes memory mintCall = getMintEncoded(config, 10e18, charlie, ZERO_BYTES);
+        calls[2] = abi.encodeWithSelector(IPositionManager.modifyLiquidities.selector, mintCall, _deadline);
+
+        uint256 tokenId = lpm.nextTokenId();
+        lpm.multicall(calls);
```
Comment (hensha256) [inline test/position-managers/PositionManager.multicall.t.sol:368]: ideally this test would check
```
bytes[] memory results = lpm.multicall(calls);
assert(results[0] is correct revert reason) 
assert(results[1] is correct revert reason) 
```

i guess that revert reason is that the nonce is taken

File: test/position-managers/PositionManager.multicall.t.sol [hunk 4]
```diff
@@ -321,4 +331,76 @@ contract PositionManagerMulticallTest is Test, Permit2SignatureHelpers, PosmTest
...
+        tokens[1] = Currency.unwrap(currency1);
+
+        IAllowanceTransfer.PermitBatch memory permit =
+            defaultERC20PermitBatchAllowance(tokens, permitAmount, permitExpiration, permitNonce);
+        permit.spender = address(lpm);
+        bytes memory sig = getPermitBatchSignature(permit, charliePK, PERMIT2_DOMAIN_SEPARATOR);
+
+        // bob front-runs the permits
+        vm.prank(bob);
+        lpm.permitBatch(charlie, permit, sig);
```
Comment (dianakocsis) [inline test/position-managers/PositionManager.multicall.t.sol:393]: stop prank

File: test/position-managers/PositionManager.multicall.t.sol [hunk 5]
```diff
@@ -321,4 +331,76 @@ contract PositionManagerMulticallTest is Test, Permit2SignatureHelpers, PosmTest
...
+        bytes[] memory calls = new bytes[](3);
+        calls[0] = abi.encodeWithSelector(Permit2Forwarder(lpm).permit.selector, charlie, permit0, sig0);
+        calls[1] = abi.encodeWithSelector(Permit2Forwarder(lpm).permit.selector, charlie, permit1, sig1);
+        bytes memory mintCall = getMintEncoded(config, 10e18, charlie, ZERO_BYTES);
+        calls[2] = abi.encodeWithSelector(IPositionManager.modifyLiquidities.selector, mintCall, _deadline);
+
+        uint256 tokenId = lpm.nextTokenId();
+        lpm.multicall(calls);
+
+        assertEq(lpm.ownerOf(tokenId), charlie);
```
Comment (dianakocsis) [inline test/position-managers/PositionManager.multicall.t.sol:370]: could you also assert that charlie was not the owner before the call? and in the other test too

**Review by dianakocsis — CHANGES_REQUESTED** [review]
looks great! just 2 things

File: test/position-managers/PositionManager.multicall.t.sol (hunk 4 above)
Comment (hensha256) [reply test/position-managers/PositionManager.multicall.t.sol:393]: you dont need to stop after `prank`
`vm.startPrank` pranks until you call `vm.stopPrank`
if you just use `vm.prank` it only pranks 1 call and theres no need to stop it

File: test/position-managers/PositionManager.multicall.t.sol (hunk 2 above)
Comment (saucepoint) [reply test/position-managers/PositionManager.multicall.t.sol:359]: mm good idea now thats try-catch'd

File: test/position-managers/PositionManager.multicall.t.sol (hunk 3 above)
Comment (saucepoint) [reply test/position-managers/PositionManager.multicall.t.sol]: can you take a look now? im trying to do the Wrap__ error handling

cc @snreynolds

**Review by snreynolds — COMMENTED** [review]
hmm to me I feel like we should allow the caller to specify whether or not they want to continue the multicall if there is an error

also if permit errors (even for a different reason) the rest of the code will still execute in this case

**Comment by saucepoint** [comment]
> hmm to me I feel like we should allow the caller to specify whether or not they want to continue the multicall if there is an error

the idea being we expose another generic `multicall(bytes calldata calls, bool revert)` where `revert = true` will cause a halt?

**Review by hensha256 — APPROVED** [review]

**Review by dianakocsis — APPROVED** [review]

**Pull request merged**